DiffusionPreprocessing/scripts/DiffusionToStructural.sh

```diff
@@ -103,15 +103,26 @@ if [ ${GdcorrectionFlag} -eq 1 ]; then
     ${FSLDIR}/bin/convertwarp --rel --relout --warp1="$DataDirectory"/warped/fullWarp --postmat="$WorkingDirectory"/diff2str.mat --ref="$T1wRestoreImage"_${DiffRes} --out="$WorkingDirectory"/grad_unwarp_diff2str
     ${FSLDIR}/bin/applywarp --rel -i "$DataDirectory"/warped/data_warped -r "$T1wRestoreImage"_${DiffRes} -w "$WorkingDirectory"/grad_unwarp_diff2str --interp=spline -o "$T1wOutputDirectory"/data
 
+    #Create a mask covering voxels within the field of view for all volumes
+    ${FSLDIR}/bin/fslmaths "$DataDirectory"/warped/data_warped -abs -Tmin -bin "$DataDirectory"/warped/fov_mask
+    ${FSLDIR}/bin/applywarp --rel -i "$DataDirectory"/warped/fov_mask -r "$T1wRestoreImage"_${DiffRes} -w "$WorkingDirectory"/grad_unwarp_diff2str --interp=trilinear -o "$T1wOutputDirectory"/fov_mask
```

Comment:
Any particular reason you are using trilinear here, rather than nearest neighbor?

Reply:
To be more conservative in which voxels to include. The voxels closest to the edge have their signal greatly reduced during the interpolation, because they are affected by the voxels outside of the FOV (with their value of zero). Nearestneighbour might include some of these voxels, while trilinear is a bit more conservative.

However, after thinking it through a bit more: we might want to actually use spline interpolate, because that was used for the data interpolation. So any voxels where the spline interpolation is less than 1, will have its signal reduced due to interpolation of the zeros outside of the FOV.  Instead of excluding all potentially affected voxels, we could do the following:
1. spline interpolate the FOV mask to T1-weighted space
2. Divide the data by the spline interpolated FOV mask (correcting for the signal reduction due to interpolation with the area outside of the FOV)
3. Afterwards we can use Nearestneighbour to exclude those voxels which were outside of the FOV (and whose diffusion MRI data is hence purely based on extrapolation due to the spline).

Reply:
So, if I understand correctly, that approach would be slightly less conservative? (i.e., would preserve a small number of extra voxels around the edges? And the division by the spline interpolated FOV mask "corrects" the signal intensity for the spline interpolation in those edge voxels that get included?)

 Seems like this may be getting more complicated than we need to be, but curious what @coalsont thinks...

Reply:
I would be wary of dividing by a spline interpolated ROI, among other things it could attempt to bring the exterior ringing fringes back in.  It "smells" like something being done wrong, and trying to find a fix that isn't doing it over the right way.

I don't entirely follow the process, but it looks like there may be two separate issues: a brain mask (for which the common advice is to only ever apply it after all your resampling is finished) and movement outside the FoV (which is missing data in the original scan, and thus should be handled differently than brain masking).  I would suggest separating these aspects before considering weird things with division.

Reply:
Thinking about it some more, the idea of dividing by a spline interpolated mask has more fundamental problems.  The ringing artifact is *additive*, not multiplicative, to the affected voxels, and its magnitude is proportional to the "cliff height" of the edge that induced it.  However, this edge size is measured locally, not globally, which makes attempting to "correct" it more troublesome.

Really, it seems like the issue is that ideally, we shouldn't be interpolating after masking out an FoV edge from other data.  Spline interpolation can reasonably be written to handle data that is nonzero at the FoV edge, but attempting to get it to ignore an edge that doesn't touch the FoV boundary is very problematic.  Dilation is the approach we have taken in similar situations.

Reply:
I am merely trying to deal with the the FoV issue here. The FoV is in effect a mask, which has been applied by the scanner before any resampling (whether due to eddy or the resampling to T1-weighted space). In eddy this has been dealty with by setting to zero any voxels that leave the FoV in any volume. This creates a sharp boundary in the eddy output data at the edge of the FoV. When resampling to T1-weighted space, this leads to ringing due to the spline interpolation. 

The question of this patch is how to deal with that ringing. I can see three solutions:
1. The aggressive solution is just to identify any voxels that are close enough to the edge of the FoV to have been affected by the sharp boundary and set them all to zero. Given the use of spline interpolation, that would be a large number of voxels.
2. The more moderate approach, which I took here is to only remove the voxels at the actual edge of the FOV which would be greatly affected. This would mean leaving the ringing caused by the resampling across the FoV's edge intact.
3. The more complex approach would be to try to correct for the ringing by dividing the spline interpolation of the FoV mask (as outlined above). However, to be honest this would be a bit of a waste of time as it multiplies all diffusion data in a voxel with the same number, which will not affect any of the diffusion MRI output (except for the S0 image).

The only interaction in this patch with the brain mask, is that I later adjust the nodif_brain_mask to exclude any voxels, where the diffusion data has been set to zero (as we don't want to run dtifit/bedpostX on those voxels).

Reply:
What I was saying is that I don't think 3 is really a solution (because it is not a multiplicative effect, among other things), and I am presenting an option 4: dilate to "invent" some data where the FoV intersects a generous brain mask, to remove the hard edges *before* resampling to T1 space.  This avoids, reduces, or at least moves the ringing artifact so that it has less effect on the real data.  It would still make sense to mask the dilated area back out after the resampling.

Ideally, eddy would output the mask without applying it to the data of the unaffected frames (which would prevent interpolation ringing in frames that didn't have large motion), or even more ideally, we would combine the motion correction and T1-space resamplings into one step somehow (which would prevent the ringing entirely when using a well-written spline interpolation).

However, option 2 sounds tolerable, and I was only arguing against the details of option 3.

Reply:
Sorry, I did not see your latest message before my previous post. Dilation before the spline resampling sounds like a much better idea than trying to correct for the ringing afterwards. I think it's worth adding to this patch and will set it up and test it in the coming days.

Reply:
wb_command's `-volume-dilate` has some options that may be useful for this purpose (specifying ROIs for what data is good and where to dilate into), and with the `WEIGHTED` method it uses a sort of smoothing kernel to generate the dilated values.  However, we have found that the defaults in 1.3.2 and earlier are not all that great, `-exponent 7` should improve the smoothness of the dilated values, and a larger dilation distance could also help with smoothness (specifying a very restricted `-bad-voxel-roi` is recommended for runtime considerations).

The next release will have better defaults, a much faster implementation, and new cutoff logic, but we don't have a timeframe planned for the release yet.